beacon-chain/db/kv/blocks.go

```diff
@@ -403,33 +403,51 @@ func (s *Store) HighestSlotBlocksBelow(ctx context.Context, slot types.Slot) ([]
 	ctx, span := trace.StartSpan(ctx, "BeaconDB.HighestSlotBlocksBelow")
 	defer span.End()
 
-	var best []byte
-	if err := s.db.View(func(tx *bolt.Tx) error {
+	var root [32]byte
+	sk := bytesutil.Uint64ToBytesBigEndian(uint64(slot))
+	err := s.db.View(func(tx *bolt.Tx) error {
 		bkt := tx.Bucket(blockSlotIndicesBucket)
-		// Iterate through the index, which is in byte sorted order.
 		c := bkt.Cursor()
-		for s, root := c.First(); s != nil; s, root = c.Next() {
+		// The documentation for Seek says:
+		// "If the key does not exist then the next key is used. If no keys follow, a nil key is returned."
+		sl, r := c.Seek(sk)
+		// So if there are slots in the index higher than the requested slot, sl will be equal to the key that is
+		// one higher than the value we want. If the slot argument is higher than the highest value in the index,
+		// we'll get a nil value for `sl`. In that case we'll go backwards from Cursor.Last().
+		if sl == nil {
+			sl, r = c.Last()
+		}
+		for {
 			if ctx.Err() != nil {
 				return ctx.Err()
 			}
-			key := bytesutil.BytesToSlotBigEndian(s)
-			if root == nil {
+			// termination condition: when .Prev() rewinds past the beginning off the collection, key will be nil.
+			// breaking without setting `root` means `root` will be the zero value, so this function will return
+			// the genesis block.
+			if sl == nil {
+				break
+			}
+			if r == nil {
```

Comment:
is it possible for there be a case where `sl` does not equal to nil but `r` does ? In that event this would be an infinite loop.

Reply:
(the code is still wrong, so I'll fix it, just letting you know the only way to trigger this bug is if we write another bug that accidentally sticks a sub-bucket into the index bucket)